File: apps/backend/src/modules/reporting/repo.ts
import { Pool } from "pg";
import type {
  DocumentType,
  ParentDocument,
} from "shared/types/modules/reporting.ts";
import {
  executeQuery,
  withCacheAsideRedis,
  withWriteThroughRedisCache,
} from "../../shared/util/query.ts";
import type { RedisClientType } from "../../db/redis-client.ts";
import {
  S3Client,
  PutObjectCommand,
  GetObjectCommand,
} from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
import { catchError } from "shared/utils/error.ts";
import type { ApiResponse } from "shared/types/response.ts";
import type { Logger } from "winston";

export interface IReportingRepo {
  getParentDocumentByType(
    userId: string,
    documentType: DocumentType,
  ): ApiResponse<ParentDocument>;
  getParentDocumentTypeByDocumentId(
    parentDocumentId: string,
  ): ApiResponse<DocumentType>;
  saveParentDocument(
    userId: string,
    documentType: DocumentType,
    filePath: string,
  ): ApiResponse<ParentDocument>;
  getParentDocumentFilePathByDocumentID(docId: string): ApiResponse<string>;
}

export class ReportingRepo implements IReportingRepo {
  private pool: Pool;
  private redisClient: RedisClientType;
  private logger: Logger;
  constructor(pool: Pool, redisClient: RedisClientType, logger: Logger) {
    this.pool = pool;
    this.redisClient = redisClient;
    this.logger = logger.child({
      service: "reporting-repo",
    });
  }

  async getParentDocumentByType(
    userId: string,
    documentType: DocumentType,
  ): ApiResponse<ParentDocument> {
    //TODO: invalidate if necessary
    const cacheKey = `parents:${userId}:documents:${documentType}`;
    return await withCacheAsideRedis(this.redisClient, cacheKey, async () => {
      const sql = `
      SELECT *
      FROM reporting.parent_documents
      WHERE user_id = $1 AND document_type = $2;
      `;
      const { data, error } = await executeQuery<ParentDocument>(
        this.pool,
        sql,
        [userId, documentType],
      );
      if (error) {
        this.logger.error(error);
        return {
          data: undefined,
          error: {
            code: 500,
            message: error.message,
          },
        };
      }
      if (data.rows.length === 0)
        return {
          data: undefined,
          error: {
            code: 404,
            message: `Document with type: ${documentType} does not exist for parent: ${userId}!`,
          },
        };
      return { data: data.rows[0], error: undefined };
    });
  }

  async getParentDocumentTypeByDocumentId(
    parentDocumentId: string,
  ): ApiResponse<DocumentType> {
    //TODO: invalidate if necessary
    const cacheKey = `parent_documents:${parentDocumentId}:type`;
    return await withCacheAsideRedis(this.redisClient, cacheKey, async () => {
      const sql = `
        SELECT document_type
        FROM reporting.parent_documents
        WHERE id = $1; 
        `;
      const { data, error } = await executeQuery<{
        document_type: DocumentType;
      }>(this.pool, sql, [parentDocumentId]);
      if (error) {
        this.logger.error(error);
        return {
          data: undefined,
          error: {
            code: 500,
            message: error.message,
          },
        };
      }
      if (data.rows.length === 0)
        return {
          data: undefined,
          error: {
            code: 404,
            message: `Parent document with id: ${parentDocumentId} was not found!`,
          },
        };
      return { data: data.rows[0].document_type, error: undefined };
    });
  }

  async saveParentDocument(
    userId: string,
    documentType: DocumentType,
    filePath: string,
  ): ApiResponse<ParentDocument> {
    const cacheKey = `parents:${userId}:parent_documents:${documentType}`;
    return await withWriteThroughRedisCache(
      this.redisClient,
      cacheKey,
      async () => {
        const sql = `
        INSERT INTO reporting.parent_documents(
          user_id,
          document_type,
          file_path)
        VALUES ($1, $2, $3)
        RETURNING *;
        `;
        const { data, error } = await executeQuery<ParentDocument>(
          this.pool,
          sql,
          [userId, documentType, filePath],
        );
        if (error) {
          this.logger.error(error);
          return {
            data: undefined,
            error: {
              code: 500,
              message: error.message,
            },
          };
        }
        if (data.rows.length === 0)
          return {
            data: undefined,
            error: {
              code: 404,
              message: `Document: ${documentType} was not saved successfully for parent: ${userId}!`,
            },
          };
        return { data: data.rows[0], error: undefined };
      },
    );
  }

  async getParentDocumentFilePathByDocumentID(
    docId: string,
  ): ApiResponse<string> {
    //TODO:invalidate on mutations
    const cacheKey = `parent_documents:${docId}`;
    return await withCacheAsideRedis(this.redisClient, cacheKey, async () => {
      const sql = `
        SELECT file_path
        FROM reporting.parent_documents
        WHERE id = $1;
        `;
      const { data, error } = await executeQuery<{ file_path: string }>(
        this.pool,
        sql,
        [docId],
      );
      if (error) {
        this.logger.error(error);
        return {
          data: undefined,
          error: {
            code: 500,
            message: error.message,
          },
        };
      }
      if (data.rows.length === 0)
        return {
          data: undefined,
          error: {
            code: 404,
            message: `Parent document with id: ${docId} was not found!`,
          },
        };
      return { data: data.rows[0].file_path, error: undefined };
    });
  }
}

export interface IS3Repository {
  uploadFile(bucket: string, key: string, file: File): ApiResponse<string>;
  getDocumentURLByFilePath(
    key?: string,
    bucket?: string,
    expiresIn?: number,
  ): ApiResponse<string>;
}

export class S3Repository implements IS3Repository {
  private client: S3Client;
  private logger: Logger;

  constructor(logger: Logger) {
    this.client = new S3Client({
      region: "us-east-1", // dummy MinIO region
      endpoint: "http://localhost:9000",
      forcePathStyle: true, // required for MinIO
      credentials: {
        accessKeyId: "minioadmin",
        secretAccessKey: "minioadmin",
      },
    });
    this.logger = logger.child({
      service: "reporting-s3-repo",
    });
  }

  async uploadFile(
    bucket: string,
    key: string,
    file: File,
  ): ApiResponse<string> {
    const arrayBuffer = await file.arrayBuffer();

    const { error } = await catchError(
      this.client.send(
        new PutObjectCommand({
          Bucket: bucket,
          Key: key,
          Body: Buffer.from(arrayBuffer),
          ContentType: file.type,
        }),
      ),
    );
    if (error) {
      this.logger.error(error);
      return {
        data: undefined,
        error: {
          code: 500,
          message: error.message,
        },
      };
    }

    return { data: `/${bucket}/${key}`, error: undefined };
  }

  async getDocumentURLByFilePath(
    key: string,
    bucket: string = "mybucket",
    expiresIn: number = 3600,
  ): ApiResponse<string> {
    const { data, error } = await catchError(
      getSignedUrl(
        this.client,
        new GetObjectCommand({
          Bucket: bucket,
          Key: key,
        }),
        {
          expiresIn,
        },
      ),
    );
    if (error) {
      this.logger.error(error);
      return {
        data: undefined,
        error: {
          code: 500,
          message: error.message,
        },
      };
    }
    return { data, error: undefined };
  }
}

File: apps/backend/src/modules/reporting/repo.py
from abc import ABC, abstractmethod
from uuid import UUID
from asyncpg import Pool
from src.shared.exceptions.not_found import NotFoundException
from src.shared.exceptions.database import DatabaseException
from src.shared.types.modules.reporting.enum import CHILD_DOCUMENT_TYPE, DOCUMENT_TYPE
from src.shared.types.modules.reporting.model import (
    ChildDocument,
    ParentDocument,
    ParentPartnerDocument,
)
from src.shared.utils.query import try_except


class IReportingRepo(ABC):
    @abstractmethod
    async def get_parent_document_by_type(
        self,
        user_id: UUID,
        document_type: DOCUMENT_TYPE,
    ) -> [ParentDocument]:
        pass

    @abstractmethod
    async def get_parent_document_file_path_by_document_id(
        self,
        doc_id: UUID,
    ) -> str:
        pass

    @abstractmethod
    async def save_parent_document(
        self,
        user_id: UUID,
        document_type: DOCUMENT_TYPE,
        file_path: str,
    ) -> ParentDocument:
        pass

    @abstractmethod
    async def get_child_document_by_type(
        self,
        child_id: UUID,
        document_type: CHILD_DOCUMENT_TYPE,
    ) -> ChildDocument:
        pass

    @abstractmethod
    async def save_child_document(
        self,
        child_id: UUID,
        document_type: CHILD_DOCUMENT_TYPE,
        file_path: str,
    ) -> ChildDocument:
        pass

    @abstractmethod
    async def get_child_document_file_path_by_document_id(
        self,
        doc_id: UUID,
    ) -> str:
        pass

    @abstractmethod
    async def get_parent_partner_document_by_type(
        self,
        partner_id: UUID,
        document_type: DOCUMENT_TYPE,
    ) -> ParentPartnerDocument:
        pass

    @abstractmethod
    async def save_parent_partner_document(
        self,
        partner_id: UUID,
        document_type: DOCUMENT_TYPE,
        file_path: str,
    ) -> ParentPartnerDocument:
        pass


class ReportingRepo(IReportingRepo):
    def __init__(self, pool: Pool) -> None:
        self.pool = pool

    async def get_parent_document_by_type(
        self,
        user_id: UUID,
        document_type: DOCUMENT_TYPE,
    ) -> ParentDocument:
        sql = """
        SELECT *
        FROM reporting.parent_documents
        WHERE user_id = $1 AND document_type = $2;
        """
        async with self.pool.acquire() as connection:
            rows, error = await try_except(
                connection.fetch, sql, user_id, document_type
            )
            if error:
                raise DatabaseException(message=str(error))
            if len(rows) == 0:
                raise NotFoundException(
                    message=f"Document with type: ${document_type} does not exist for parent: {user_id}!",
                )
            return ParentDocument(**rows[0])

    async def get_parent_document_file_path_by_document_id(
        self,
        doc_id: UUID,
    ) -> str:
        sql = """
        SELECT file_path
        FROM reporting.parent_documents
        WHERE id = $1;
        """
        async with self.pool.acquire() as connection:
            rows, error = await try_except(connection.fetch, sql, doc_id)
            if error:
                raise DatabaseException(message=str(error))
            if len(rows) == 0:
                raise NotFoundException(
                    message=f"Parent document with id: {doc_id} was not found!",
                )
            return rows[0]["file_path"]

    async def save_parent_document(
        self,
        user_id: UUID,
        document_type: DOCUMENT_TYPE,
        file_path: str,
    ) -> ParentDocument:
        sql = """
        INSERT INTO reporting.parent_documents(
          user_id,
          document_type,
          file_path)
        VALUES ($1, $2, $3)
        RETURNING *;
        """
        async with self.pool.acquire() as connection:
            row, error = await try_except(
                connection.fetchrow, sql, user_id, document_type, file_path
            )
            if error:
                raise DatabaseException(message=str(error))
            if row is None:
                raise NotFoundException(
                    message=f"Document: ${document_type} was not saved successfully for parent: ${user_id}!",
                )
            return ParentDocument(**row)

    async def get_child_document_by_type(
        self,
        child_id: UUID,
        document_type: CHILD_DOCUMENT_TYPE,
    ) -> ChildDocument:
        sql = """
        SELECT *
        FROM reporting.children_documents
        WHERE child_id = $1 AND document_type = $2;
        """
        async with self.pool.acquire() as connection:
            rows, error = await try_except(
                connection.fetch, sql, child_id, document_type
            )
            if error:
                raise DatabaseException(code=500, message=str(error))
            if len(rows) == 0:
                raise NotFoundException(
                    message=f"Document with type: ${document_type} does not exist for child: {child_id}!",
                )
            return ChildDocument(**rows[0])

    async def save_child_document(
        self,
        child_id: UUID,
        document_type: CHILD_DOCUMENT_TYPE,
        file_path: str,
    ) -> ChildDocument:
        sql = """
        INSERT INTO reporting.children_documents(
          child_id,
          document_type,
          file_path)
        VALUES ($1, $2, $3)
        RETURNING *;
        """
        async with self.pool.acquire() as connection:
            row, error = await try_except(
                connection.fetchrow, sql, child_id, document_type, file_path
            )
            if error:
                raise DatabaseException(message=str(error))
            if row is None:
                raise NotFoundException(
                    message=f"Document: ${document_type} was not saved successfully for child: ${child_id}!",
                )
            return ChildDocument(**row)

    async def get_child_document_file_path_by_document_id(
        self,
        doc_id: UUID,
    ) -> str:
        sql = """
        SELECT file_path
        FROM reporting.children_documents
        WHERE id = $1;
        """
        async with self.pool.acquire() as connection:
            rows, error = await try_except(connection.fetch, sql, doc_id)
            if error:
                raise DatabaseException(message=str(error))
            if len(rows) == 0:
                raise NotFoundException(
                    message=f"Child document with id: {doc_id} was not found!",
                )
            return rows[0]["file_path"]

    async def get_parent_partner_document_by_type(
        self,
        partner_id: UUID,
        document_type: DOCUMENT_TYPE,
    ) -> ParentPartnerDocument:
        sql = """
        SELECT *
        FROM reporting.parent_partner_documents
        WHERE partner_id = $1 AND document_type = $2;
        """
        async with self.pool.acquire() as connection:
            rows, error = await try_except(
                connection.fetch, sql, partner_id, document_type
            )
            if error:
                raise DatabaseException(message=str(error))
            if len(rows) == 0:
                raise NotFoundException(
                    message=f"Document with type: ${document_type} does not exist for parent partner: {partner_id}!",
                )
            return ParentPartnerDocument(**rows[0])

    async def save_parent_partner_document(
        self,
        partner_id: UUID,
        document_type: DOCUMENT_TYPE,
        file_path: str,
    ) -> ParentPartnerDocument:
        sql = """
        INSERT INTO reporting.parent_partner_documents(
          partner_id,
          document_type,
          file_path)
        VALUES ($1, $2, $3)
        RETURNING *;
        """
        async with self.pool.acquire() as connection:
            row, error = await try_except(
                connection.fetchrow, sql, partner_id, document_type, file_path
            )
            if error:
                raise DatabaseException(message=str(error))
            if row is None:
                raise NotFoundException(
                    message=f"Document: ${document_type} was NOT saved successfully for parent partner: ${partner_id}!",
                )
            return ParentPartnerDocument(**row)
